Add english version of CMS docs (#4272)
ydb-platform/ydb: ydb/docs/en/core/maintenance/manual/maintenance-without-downtime.md

{% endnote %}

## Maintenance task {#maintenance-task}

A *maintenance task* is a set of *actions* that the user asks the CMS to perform for safe maintenance.

Supported actions:
- Acquiring an exclusive lock on a cluster component — node or host.

In a task, actions are divided into groups. Actions from the same group are performed atomically. Currently, groups can consist of only one action.

If it's not possible to perform an action at the time of the request, the CMS informs you of the reason and the time when it is worth *refreshing* the task, and sets the action status to *pending*. When the task is refreshed, the CMS attempts to perform the pending actions again.

*Performed* actions have a deadline after which they are considered *completed* and stop having an effect on the cluster. For example, an exclusive lock is released. An action can be completed early.

{% note info "Protracted maintenance" %}

If maintenance continues after the actions that were performed to make it safe have been completed, this is considered a fault in the cluster.

{% endnote %}

Completed actions are automatically removed from the task.

> ### Availability mode {#availability-mode}
> 
> In a maintenance task, you need to specify the availability mode of the cluster to be complied when checking whether actions can be performed. The following modes are supported:
> - **Strong** - a mode that minimizes the risk of availability loss.
>     - No more than one unavailable [VDisk](../../concepts/cluster/distributed_storage.md#storage-groups) is allowed in each affected storage group.
>     - No more than one unavailable State Storage ring is allowed.
> - **Weak** - a mode that does not allow exceeding the failure model.
>     - No more than two unavailable VDisks are allowed for affected storage groups with the [block-4-2](../../deploy/configuration/config.md#reliability) scheme.
>     - No more than four unavailable VDisks, three of which must be in the same data center, are allowed for affected storage groups with the [mirror-3-dc](../../deploy/configuration/config.md#reliability) scheme. 
>     - No more than `(nto_select - 1) / 2` unavailable State Storage rings are allowed.
> - **Force** - forced mode, the failure model is ignored. Not recommended for use.

Reviewer: ```suggestion
### Availability mode {#availability-mode}

In a maintenance task, you need to specify the cluster's availability mode to comply with when checking whether actions can be performed. The following modes are supported:
- **Strong**: a mode that minimizes the risk of availability loss.
    - No more than one unavailable [VDisk](../../concepts/cluster/distributed_storage.md#storage-groups) is allowed in each affected storage group.
    - No more than one unavailable State Storage ring is allowed.
- **Weak**: a mode that does not allow exceeding the failure model.
    - For affected storage groups with the [block-4-2](../../deploy/configuration/config.md#reliability) scheme, no more than two unavailable VDisks are allowed.
    - For affected storage groups with the [mirror-3-dc](../../deploy/configuration/config.md#reliability) scheme, up to four unavailable VDisks are allowed, three of which must be in the same data center. 
    - No more than `(nto_select - 1) / 2` unavailable State Storage rings are allowed.
- **Force**: a forced mode, the failure model is ignored. *Not recommended for use.*
```

> ### Priority {#priority}
> 
> You can specify the priority of a maintenance task. A lower value means a higher priority.
> 
> The actions of the task cannot be performed until all conflicting actions from tasks with a higher priority are completed. Tasks with the same priority have no advantage over each other.

Reviewer: ```suggestion
### Priority {#priority}

You can specify the priority of a maintenance task. A lower value means a higher priority.

The task's actions cannot be performed until all conflicting actions from tasks with a higher priority are completed. Tasks with the same priority have no advantage over each other.
```

    - Whether it's possible to lock the node according to the limits of unavailable nodes.
    - Whether it's possible to lock all VDisks of the node according to the availability mode.
    - Whether it's possible to lock the State Storage ring of the node according to the availability mode.
    - Whether it's possible to lock the node according to the limit of unavailable nodes on which cluster system tablets can run.

If the checks are successful, the action can be performed and a temporary locks are acquired on the checked nodes. The CMS then considers the next group of actions. Temporary locks help to understand whether the actions requested in different groups conflict with each other. Once the check is complete, the temporary locks are released.

## Examples {#examples}

The [ydbops](https://github.com/ydb-platform/ydbops) utility tool uses CMS for cluster maintenance without downtime. You can also use the CMS directly through the [gRPC API](https://github.com/ydb-platform/ydb/blob/main/ydb/public/api/grpc/draft/ydb_maintenance_v1.proto).

### Take out a node for maintenance {#node-maintenance}

{% note info "Functionality in development" %}

> Functionality is expected in upcoming versions of ydbops.

Reviewer: ```suggestion
Functionality is expected in upcoming versions of `ydbops`.
```

{% endnote %}

To take out a node for maintenance, you can use the command:
```
$ ydbops node maintenance --host <node_fqdn>
```

> When executing this command, ydbops will acquire an exclusive lock on the node in CMS.

Reviewer: ```suggestion
When executing this command, `ydbops` will acquire an exclusive lock on the node in CMS.
```

> When executing this command, ydbops will acquire an exclusive lock on the node in CMS.

Reviewer: This paragraph sounds weird: typically, multiple YDB nodes are on a single host, and such a command probably needs to acquire locks for all of them. Or maybe this implies only static/storage nodes, but this would need to be explicitly clarified, too.

Author: Changed parameter to `<node_id>` to make it more clear. I just took this command from https://github.com/ydb-platform/ydbops/issues/2
It is under active development. When the final variant of the command is chosen, the text in this article will be changed.

Reviewer: @pixcc I don't think this change fixes the problem: in this article, we **shouldn't** use "node" and "host" interchangeably. This command sounds more like a maintenance of all (or some) nodes on a given host but not of a single node. Meanwhile, the surrounding text sounds like it is about a single node.

Author: I removed command to eliminate ambiguity.

It's not 100% clear whether the final version of the command will take out a host for maintenance (with all nodes) or individual nodes, so I've settled on a simple node case for now.

> ### Rolling restart {##rolling-restart}
> 
> To perform a rolling restart of the entire cluster you can use the command:

Reviewer: ```suggestion
### Rolling restart {##rolling-restart}

To perform a rolling restart of the entire cluster, you can use the command:
```

```

> $ ydbops restart --endpoint grpc://<cluster-fqdn> --availability-mode strong

Reviewer: Doesn't this have some additional requirements? For instance, the last time I tried (a while ago), it couldn't restart a cluster deployed with Ansible because it relied on some different systemd unit naming.

Reviewer 2: Things have changed a bit. By default the `ydb-server-storage.service` systemd name will be used, but it is possible to specify a different systemd unit name with `--systemd-unit` flag (here is a relevant `--help` output which contains this flag: https://pastebin.com/Jqcx31Jn). 

Just for context: the fact that we have two different default unit names for deploying with Ansible and for everything else (cloud environments etc.) - is horrible and we just have to live with it for a while. 

It is probably a good idea to include the mention of `--system-unit` flag into the docs, but not in too much detail, maybe one sentence only. E.g. "If your systemd unit name is different from the default one, you may need to override it with --systemd-unit flag"

Author: Added aditional requirements

```

> The ydbops utility will automatically create a maintenance task to restart the entire cluster using the given availability mode. As it progresses, ydbops will refresh the maintenance task and acquire exclusive locks on the nodes in the CMS until all nodes are restarted.

Reviewer: ```suggestion
The `ydbops` utility will automatically create a maintenance task to restart the entire cluster using the given availability mode. As it progresses, `ydbops` will refresh the maintenance task and acquire exclusive locks on the nodes in the CMS until all nodes are restarted.
```
